sql/core/src/main/java/org/apache/spark/sql/sources/v2/ReadMicroBatchSupport.java

```diff
@@ -0,0 +1,35 @@
+package org.apache.spark.sql.sources.v2;
+
+import java.util.Optional;
+
+import org.apache.spark.annotation.InterfaceStability;
+import org.apache.spark.sql.execution.streaming.Offset;
+import org.apache.spark.sql.sources.v2.reader.DataSourceV2Reader;
+
+/**
+ * A mix-in interface for {@link DataSourceV2}. Data sources can implement this interface to
+ * provide streaming microbatch data reading ability.
+ */
+@InterfaceStability.Evolving
+public interface ReadMicroBatchSupport extends BaseStreamingSource {
```

Comment:
I ended up changing the interface to setOffsetRange(Optional<Offset> start, Offset end) for consistency with the other mixins, and to avoid having to pass offsets outside the StreamExecution.

With this interface, I don't think the limitation is nearly as awkward; it makes sense that we'd have to call setOffsetRange first if sources are allowed to implement getOffset() in a way that depends on the "current" offset. An init method would probably end up having the same signature and doing the same thing.

I agree it would be nicer if sources could provide rate limiting without maintaining complex internal state like this, but I think the solution there looks like a new API with incompatible semantics. (Maybe something like setOffsetRange(Optional<Offset> start, long estNumBytes).) I don't see a way to fit maxOffsetsPerTrigger, maxFilesPerBatch, all the Kinesis RateLimits, etc. into a single extensible API.